velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -549,28 +550,41 @@ std::string toString(const MaterializedRow& row) {
   return oss.str();
 }
 
-std::string generateUserFriendlyDiff(
-    const MaterializedRowMultiset& expectedRows,
-    const MaterializedRowMultiset& actualRows) {
-  std::vector<MaterializedRow> extraActualRows;
-  std::set_difference(
-      actualRows.begin(),
-      actualRows.end(),
-      expectedRows.begin(),
-      expectedRows.end(),
-      std::inserter(extraActualRows, extraActualRows.end()));
+bool lessThanWithEpsilon(
+    const MaterializedRow& lhs,
+    const MaterializedRow& rhs) {
+  const auto minSize = std::min(lhs.size(), rhs.size());
+  for (size_t i = 0; i < minSize; ++i) {
+    if (lhs[i].equalsWithEpsilon(rhs[i])) {
+      continue;
+    }
+    // The 1st non-equal element determines if 'left' is smaller or not.
+    return lhs[i].lessThanWithEpsilon(rhs[i]);
+  }
+  // The shorter vector is smaller.
+  return lhs.size() < rhs.size();
+}
 
-  std::vector<MaterializedRow> missingActualRows;
-  std::set_difference(
-      expectedRows.begin(),
-      expectedRows.end(),
-      actualRows.begin(),
-      actualRows.end(),
-      std::inserter(missingActualRows, missingActualRows.end()));
+bool equalWithEpsilon(const MaterializedRow& lhs, const MaterializedRow& rhs) {
+  if (lhs.size() != rhs.size()) {
+    return false;
+  }
+  auto size = lhs.size();
+  for (auto i = 0; i < size; ++i) {
+    if (!lhs[i].equalsWithEpsilon(rhs[i])) {
+      return false;
+    }
+  }
+  return true;
+}
 
+std::string makeErrorMessage(
+    const std::vector<MaterializedRow>& missingActualRows,
```

Comment:
nit: consider renaming to missingRows and actualRows. missingActualRows doesn't quite make sense, because these rows are missing from the 'actual' results.